Added mongodb atlas mapstore-maploader tutorial (#1)
hazelcast-guides/hazelcast-platform-operator-map-store: docs/modules/ROOT/pages/hazelcast-platform-operator-map-store-mongodb-atlas.adoc

> = Configure MongoDb Atlas as an External Data Store for the Cluster with Hazelcast Platform Operator

Reviewer: ```suggestion
= Configure MongoDB Atlas as an External Data Store for the Cluster with Hazelcast Platform Operator
```

:page-layout: tutorial
:page-product: operator

> :page-categories: Cloud Native

Reviewer: ```suggestion
:page-categories: Kubernetes
```

:page-lang: java
:page-enterprise: false
:page-est-time: 20 mins
:description: Learn how to keep data in sync across two Hazelcast clusters.

== Context
In this tutorial, you'll do the following:

> - Build and Deploy custom MapStore/MapLoader implementation to cloud storage.

Reviewer: ```suggestion
- Build and deploy a custom MapStore/MapLoader implementation to cloud storage.
```

- Start Hazelcast cluster with classpath that includes MapStore/MapLoader JAR

- Create a Hazelcast Map with MapStore/MapLoader configuration

> - Run Hazelcast Java client to manipulate data via populating all entries from Atlas DB.

Reviewer: ```suggestion
- Run Hazelcast Java client to manipulate data via populating all entries from MongoDB Atlas.
```

== Before you Begin

Before starting this tutorial, make sure that you meet the following prerequisites:

* Up and running *public* https://kubernetes.io/[Kubernetes] cluster. Make sure the Kubernetes cluster is public.
* Kubernetes command-line tool, https://kubernetes.io/docs/tasks/tools/#kubectl[kubectl]
* Maven command-line tool, https://maven.apache.org/download.cgi[mvn]

> * MongoDB Atlas free-tier database, https://www.mongodb.com/docs/atlas/getting-started[Get Started with Atlas]

Reviewer 2: ```suggestion
* MongoDB Atlas database, https://www.mongodb.com/docs/atlas/getting-started[Get Started with Atlas] (free-tire is suitable for completing this guide).
```
Maybe the wording can be different, but my point is that the prerequisite is MongoDB Atlas DB, not exactly free-tier DB 😃

* Created blob storage and access credentials in one of the cloud providers: https://aws.amazon.com/s3/[AWS] - https://cloud.google.com/storage/[GCP] - https://azure.microsoft.com/en-us/services/storage/blobs/[Azure]

> == Step 0. Add Node IPs in MongoDB Atlas whitelist 

Reviewer: We should start from step 1.

> The database should be reachable from Hazelcast pods. That's why public IP(s) of the node(s) that they are running on should be whitelisted.  

Reviewer: ```suggestion
Hazelcast pods need access to the database. To allow Hazelcast pods to connect to the database you need to add the public IP addresses of the Kubernetes nodes to the whitelist in the Atlas dashboard.
```

NAME       STATUS        EXTERNAL-IP     
gke-....   Ready   ...   34...           ...
gke-....   Ready   ...   35...           ...
gke-....   Ready   ...   34...           ...
----

. Add IPs into whitelist at Atlas dashboard:

image::mongo-atlas-network-access.png[]

== Step 1. Deploy MapStore/MapLoader JAR to external storage

. Build MapStore/MapLoader JAR 

+
[source, shell]
----
$ cd mapstore
$ mvn package
----

. Upload the JAR to the cloud provider storage blob/bucket that you have already created:

+
[tabs]
====

S3::
+
--
[source,bash]
----
$ aws s3 cp mapstore/target/mapstore-mongodb-1.0-SNAPSHOT.jar s3://<BUCKET_NAME>
----
--

GCS::
+
--
[source,bash]
----
$ gsutil cp mapstore/target/mapstore-mongodb-1.0-SNAPSHOT.jar gs://<BUCKET_NAME>     
----
--

ASB::
+
--
[source,bash]
----
$ az storage blob upload --account-name <ACCOUNT_NAME> --container-name <CONTAINER_NAME> --file mapstore/target/mapstore-mongodb-1.0-SNAPSHOT.jar
----
--

====

== Step 2. Start Hazelcast Cluster

. Create Secret for Bucket credentials
+
Run one of the following command to create the secret according to the cloud provider you want to backup.

+
[tabs]
====

S3::
+
--
[source,bash]
----

> $ kubectl create secret generic <secret-name> --from-literal=region=<region> \

Reviewer: It's best not to include the `$`. This will get copied into the clipboard along with the command.

===============================

NOTE: Exposing the cluster for unisocket clients is required for local client that you will run in next steps. 

== Step 4. Create Hazelcast Map with MapStore/MapLoader configuration

. Create secret for database properties 
+
[source, shell]
----
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: mongodb-atlas-properties
stringData:
  mongo.url: "MONGODB_URL"
  mongo.db: "DB_NAME"
  mongo.collection: "supplements"
EOF
----
+

TIP: Sample MongoDB Atlas URL -> mongodb+srv://<USERNAME>:<PASSWORD>@<DB_NAME>.mongodb.net/?retryWrites=true&w=majority

. Create a Map with MapStore configuration:
+
[source, shell]
----
cat <<EOF | kubectl apply -f -
apiVersion: hazelcast.com/v1alpha1
kind: Map
metadata:
  name: supplements
spec:
  hazelcastResourceName: hazelcast
  mapStore:
    className: com.operator.tutorial.mongodb.MongoMapStore
    propertiesSecretName: mongodb-atlas-properties
EOF
----
+

== Step 5. Run local Hazelcast client to manipulate data

. Check external IP of the LoadBalancer:
+
[source, shell]
----
kubectl get services                           
NAME         TYPE           EXTERNAL-IP   PORT(S)

> hazelcast    LoadBalancer   34....   5701:30164/TCP

Reviewer 2: Maybe it's better to use the IP address from the range reserved from documentation, just for clarity? (e.g. `233.252.0.0-233.252.0.255`)

Author: This one is more confusing, I think 🤔 
```suggestion
hazelcast    LoadBalancer   233.252.0.0-233.252.0.255   5701:30164/TCP
```

Reviewer 2: I've meant just using one address from the range, e.g., `233.252.0.34`. But it's just a small nit, so we can skip it
